### PR TITLE
tests if custom job model methods work

### PR DESCRIPTION
**Pull request by ksiedlarek**

### CODE REVIEW

File: jobs/tests.py [hunk 1]
```diff
@@ -1,3 +1,28 @@
...
+	def setUp(self):
+		self.job = mommy.make(Job, ready_to_publish=True)
+
+	def test_is_published(self):
+		published_date_is_not_none = False
+		self.job.publish()
+		if self.job.published_date:
+			published_date_is_not_none = True
+		self.assertEqual(published_date_is_not_none, True)
+		print "Published date:{0}".format(self.job.published_date)
```
Comment (rebkwok) [inline jobs/tests.py]: This test can be simplified quite a lot:

```
self.assertFalse(self.job.published_date)
self.job.publish()
self.assertTrue(self.job.published_date, "Job has no published date")
```

So we assert that there is no published date first, then publish, then assert that there now is a published date. (When we get to writing the views tests, we can also check that the view now displays the published test)
You can use the optional last argument in the assert to print a message if the test fails.  We don't really care too much about the actual publish date (except for debugging while writing the test)

File: jobs/tests.py [hunk 2]
```diff
@@ -1,3 +1,28 @@
...
+
+from jobs.models import Job
+
+
+class JobMethodTests(TestCase):
+
+	def setUp(self):
+		self.job = mommy.make(Job, ready_to_publish=True)
+
+	def test_is_published(self):
```
Comment (rebkwok) [inline jobs/tests.py]: Add a doc string to describe what this test is doing (e.g. "Test that a job marked ready to publish has a published_date after publishing)

File: jobs/tests.py [hunk 3]
```diff
@@ -1,3 +1,28 @@
 from django.test import TestCase
 
-# Create your tests here.
+from model_mommy import mommy
+
+from jobs.models import Job
+
+
+class JobMethodTests(TestCase):
```
Comment (rebkwok) [inline jobs/tests.py]: Suggest "JobModelTests" (since later we will have View tests and Form tests etc, which may also have methods)

File: jobs/tests.py (hunk 1 above)
Comment (rebkwok) [inline jobs/tests.py]: I'd suggest that you test the negative case too - create a job that has ready_to_publish False, and test that trying to publish it doesn't succeed.
Also test that what happens if you try to publish a job that's already published (with ready_to_publish True and False)

File: jobs/tests.py [hunk 4]
```diff
@@ -1,3 +1,28 @@
...
+		print "Published date:{0}".format(self.job.published_date)
+
+	def test_expiration_date(self):
+		expiration_date_is_not_none = False
+		self.job.publish()
+		self.job.set_expiration_date()
+		if self.job.expiration_date:
+			expiration_date_is_not_none = True
+		self.assertEqual(expiration_date_is_not_none, True)
+		print "Expiration date:{0}".format(self.job.expiration_date)
```
Comment (rebkwok) [inline jobs/tests.py]: Could be simplified similarly:

```
self.job.publish()
self.assertFalse(self.job.expiration_date)
self.job.set_expiration_date()
self.assertTrue(self.job.expiration_date, "Job has no expiration date")
```

You could also check that the expiration date is what you expect (find self.job.expiration_date against timezone.now().  This gives you a timedelta object that will be in days, mins and secs - do .days on the timedelta object to get the number of days - it should be 59

Comment (rebkwok) [inline jobs/tests.py]: Also check other edge cases that you can think of - what happens if you try to set an expiration date when there's one already set, what if you try to set an expiration date on a job that's not ready to publish?  You can also write tests for the cases that we might need in the future - e.g. can you publish without having reviewed?  At the moment we can - but we might want to change that behaviour later.

File: jobs/tests.py [hunk 5]
```diff
@@ -1,3 +1,28 @@
...
+
+	def test_is_published(self):
+		published_date_is_not_none = False
+		self.job.publish()
+		if self.job.published_date:
+			published_date_is_not_none = True
+		self.assertEqual(published_date_is_not_none, True)
+		print "Published date:{0}".format(self.job.published_date)
+
+	def test_expiration_date(self):
```
Comment (rebkwok) [inline jobs/tests.py]: add docstring

File: jobs/tests.py [hunk 6]
```diff
@@ -1,3 +1,28 @@
 from django.test import TestCase
 
-# Create your tests here.
+from model_mommy import mommy
```
Comment (rebkwok) [inline jobs/tests.py:9]: add model_mommy to requirements.txt

**Comment by rebkwok** [comment]
Hi Kasia
I've added a bunch of comments on the tests and also some suggestions for others - basically, test every odd case you can think of that might happen if people try to review/publish etc in the order we didn't intend.  Then when the models change or new methods are written, we can run all the tests and check if the new things inadvertently break something else.
(There's no rush on this though - the tests aren't going to hold up anything else)
Becky

**Comment by ksiedlarek** [comment]
ok, I've added some tests, and changed a bit those which exist. I also changed set_expiration_date, but I am not sure if this was the correct way of doing things..

File: jobs/models.py [hunk 7]
```diff
@@ -60,9 +60,12 @@ def publish(self):
...
     def set_expiration_date(self):
-        if self.ready_to_publish and not self.expiration_date:
-            self.expiration_date = self.published_date + timedelta(60)
-            self.save()
+        try:
+            if self.ready_to_publish and not self.expiration_date:
+                self.expiration_date = self.published_date + timedelta(60)
+                self.save()
+        except TypeError:
+            return "Error"
```
Comment (rebkwok) [inline jobs/models.py]: You get a type error if you do set_expiration_date and there's no published_date, right?  This would be better covered in the if statement, since it's a known case

File: jobs/tests.py [hunk 8]
```diff
@@ -1,3 +1,82 @@
...
+
+from jobs.models import Job
+
+
+class JobModelTests(TestCase):
+    """Tests for custom methods of Job model."""
+
+    """Setting up models with mommy."""
```
Comment (rebkwok) [inline jobs/tests.py]: move all docstrings to the line below the function definition (just the standard convention)

File: jobs/tests.py [hunk 9]
```diff
@@ -1,3 +1,82 @@
...
+    def test_publish_unsuccesfully(self):
+        self.assertFalse(self.job1.published_date)
+        self.job1.publish()
+        self.assertFalse(self.job1.published_date, "Job has published date.")
+
+    """Publish once, than change ready_to_publish field value and publish once more."""
+    def test_publish_more_than_once_and_change_values(self):
+        self.assertFalse(self.job.published_date)
+        self.job.publish()
+        self.job.ready_to_publish == False
```
Comment (rebkwok) [inline jobs/tests.py]: to change the value, use
self.ready_to_publish = False
(self.ready_to_publish == False is just a statement)

File: jobs/tests.py [hunk 10]
```diff
@@ -1,3 +1,82 @@
...
+        # Right now if we change ready_to_publish to False, published_date stays as it was and doesn't become None.
```
Comment (rebkwok) [inline jobs/tests.py]: is this what we want to happen?  (might need some discussion about how we expect the ready_to_publish and published_date  to work - once a job is published, do we want the published date to stay the same, or do we want to be able to re-publish?  ).

File: jobs/tests.py [hunk 11]
```diff
@@ -1,3 +1,82 @@
...
+        self.job.ready_to_publish == True
+        self.job.publish()
+        self.assertTrue(self.job.published_date, "Job has no published date.")
+
+    """Publish twice in a row."""
+    def test_publish_twice_in_a_row(self):
+        self.assertFalse(self.job.published_date)
+        self.job.publish()
+        self.job.publish()
+        self.assertTrue(self.job.published_date, "Job has no published date.")
```
Comment (rebkwok) [inline jobs/tests.py:55]: similar comment to one on previous test - at the moment (I think) publishing twice will give you a new publish date each time (you could check this by assigning published_date to a variable each time and using an assert that compares the two).  Do we want to allow this, or do we want the publish() method to do something different if it already has a publish date?  If we care about it, maybe we need a "last_published_date" so we record the initial published date and the date of the last re-publish?

File: jobs/tests.py [hunk 12]
```diff
@@ -1,3 +1,82 @@
...
+        self.assertFalse(self.job.published_date)
+        self.job.publish()
+        self.job.ready_to_publish == False
+        # Right now if we change ready_to_publish to False, published_date stays as it was and doesn't become None.
+        self.assertTrue(self.job.published_date)
+        self.job.ready_to_publish == True
+        self.job.publish()
+        self.assertTrue(self.job.published_date, "Job has no published date.")
+
+    """Publish twice in a row."""
```
Comment (rebkwok) [inline jobs/tests.py]: Expand comment a bit to describe what we're testing happens when we publish twice in a row

File: jobs/tests.py [hunk 13]
```diff
@@ -1,3 +1,82 @@
...
+    def test_publish_twice_in_a_row(self):
+        self.assertFalse(self.job.published_date)
+        self.job.publish()
+        self.job.publish()
+        self.assertTrue(self.job.published_date, "Job has no published date.")
+
+    """Publish without review."""
```
Comment (rebkwok) [inline jobs/tests.py]: Make comment a bit more descriptive.  ALso, at the moment we can publish without review - if we don't want this to happen, or if we need to decide if it can happen, add a #TODO comment (in most text editors it'll be highlighted and will remind us something needs to be looked at for this test

File: jobs/tests.py [hunk 14]
```diff
@@ -1,3 +1,82 @@
...
+    def test_publish_without_review(self):
+        self.assertFalse(self.job.review_status, "Job was already reviewed.")
+        self.job.publish()
+        self.assertTrue(self.job.published_date, "Job has no published date.")
+
+
+    """Test if default value is set."""
+    def test_set_expiration_date(self):
```
Comment (rebkwok) [inline jobs/tests.py]: this one is redundant, since the next test tests both that a default is set, and what it's set to

File: jobs/tests.py [hunk 15]
```diff
@@ -1,3 +1,82 @@
...
+        self.job.publish()
+        self.assertFalse(self.job.expiration_date)
+        self.job.set_expiration_date()
+        result = self.job.expiration_date - timezone.now()
+        self.assertEqual(result.days, 59)
+
+    """Test setting expiration date if job is unpublished."""
+    def test_expiration_date_on_not_ready(self):
+        value = self.job.set_expiration_date()
+        self.assertEqual(value, "Error")
```
Comment (rebkwok) [inline jobs/tests.py]: See comments on the models file.  If a job is unpublished, we just shouldn't be able to set an expiration date. We don't want to be returning a known error.

File: jobs/tests.py [hunk 16]
```diff
@@ -1,3 +1,82 @@
...
+        self.assertFalse(self.job.expiration_date)
+
+    """Test setting expiration date more than once in a row."""
+    def test_set_expiration_date_twice(self):
+        self.job.publish()
+        self.assertFalse(self.job.expiration_date)
+        self.job.set_expiration_date()
+        self.assertTrue(self.job.expiration_date, "Job has no expiration date.")
+        self.job.set_expiration_date()
+        self.assertTrue(self.job.expiration_date, "Job has no expiration date.")
```
Comment (rebkwok) [inline jobs/tests.py:91]: might want to check the date is changing - i.e. we can extend it by reapplying set_exp_date (should just be able to check that the second date is > the first, or that one minus the other is > 0)

File: jobs/tests.py [hunk 17]
```diff
@@ -1,3 +1,82 @@
...
+        self.job.publish()
+        self.assertTrue(self.job.published_date, "Job has no published date.")
+
+    """Publish doesn't work when ready_to_publish field is set to False."""
+    def test_publish_unsuccesfully(self):
+        self.assertFalse(self.job1.published_date)
+        self.job1.publish()
+        self.assertFalse(self.job1.published_date, "Job has published date.")
+
+    """Publish once, than change ready_to_publish field value and publish once more."""
```
Comment (rebkwok) [inline jobs/tests.py]: line too long for PEP8 conventions

File: jobs/tests.py (hunk 10 above)
Comment (rebkwok) [inline jobs/tests.py]: line too long for PEP8 conventions

File: jobs/models.py (hunk 7 above)
Comment (ksiedlarek) [inline jobs/models.py]: yes, I've changed it.

File: jobs/tests.py (hunk 8 above)
Comment (ksiedlarek) [inline jobs/tests.py]: done.

File: jobs/tests.py (hunk 9 above)
Comment (ksiedlarek) [inline jobs/tests.py]: right, done.

File: jobs/tests.py (hunk 10 above)
Comment (ksiedlarek) [inline jobs/tests.py]: I am not sure if we want this to happen.. if we republish than date changes. If old date stays until republish we know that it was already published and taken down, it may be good knowledge to have, but also it may be not really useful.

File: jobs/tests.py (hunk 11 above)
Comment (ksiedlarek) [inline jobs/tests.py:55]: if we had last published date maybe we should just keep track of all what is happening to the job? I mean all of the times something was published and unpublished. But maybe this is too much? Will people really change their minds often? ;)

File: jobs/tests.py (hunk 12 above)
Comment (ksiedlarek) [inline jobs/tests.py]: ok.

File: jobs/tests.py (hunk 13 above)
Comment (ksiedlarek) [inline jobs/tests.py]: done.

File: jobs/tests.py (hunk 14 above)
Comment (ksiedlarek) [inline jobs/tests.py]: I removed it

File: jobs/tests.py (hunk 15 above)
Comment (ksiedlarek) [inline jobs/tests.py]: ok, is it good now?

File: jobs/tests.py (hunk 16 above)
Comment (ksiedlarek) [inline jobs/tests.py:91]: it is not - I think I have to republish to get new expiration date.

File: jobs/tests.py (hunk 17 above)
Comment (ksiedlarek) [inline jobs/tests.py]: ok, corrected

**Comment by ksiedlarek** [comment]
If I don't set expiration date to None, than it doesn't work, dates are the same. But I don't have to set published date to None, so I've deleted it.

**Comment by rebkwok** [comment]
Is this a bug in the code?  Should be being set (i.e. extended) each time? Don't set the expiration date back to None in the test to make the test work if it's not something that happens in the code (unless we're expecting users to delete the expiration date before they try to set it again)

**Comment by ksiedlarek** [comment]
is it better now?

File: jobs/tests.py [hunk 18]
```diff
@@ -1,3 +1,92 @@
...
+        self.assertFalse(self.job.expiration_date)
+        self.job.set_expiration_date()
+        first_date = self.job.expiration_date
+        self.assertTrue(self.job.expiration_date, "Job has no expiration date.")
+        time.sleep(1)
+        self.job.publish()
+        self.job.set_expiration_date()
+        second_date = self.job.expiration_date
+        self.assertTrue(self.job.expiration_date, "Job has no expiration date.")
+        self.assertTrue(second_date.second - first_date.second > 0)
```
Comment (rebkwok) [inline jobs/tests.py]: we can just use

```
self.assertTrue(second_date > first_date)
```

Comment (ksiedlarek) [inline jobs/tests.py]: ok,done.
